### ISSUE TITLE
Nuanced types

### ISSUE THREAD

**Issue by oscartbeaumont** [specta-rs/specta#203]
Specta needs some way of tracking types at a lower level to handle special Typescript types.

This needs to support:
 - [ ] `BigInt`
 - [ ] `Date`
 - [ ] `UInt8Array` for [`serde_bytes`](https://docs.rs/serde_bytes) w/ opt-in for `Vec<u8>` and opt-out

These types break down into two categories.
 - Where serialization is loseless (Eg. `Date`, `UInt8Array`) so the decoding just needs to map the existing value into a different type in JS
 - Lossy types like BigInt's or `Nan`/`Infinity` which get truncated or become `null` before reaching usercode in the webview.

Currently loseless types are implemented in `specta-tags` although the whole crate is very experimental at the moment. The focus now is on figuring out lossy types.

For the lossy types we need runtime support in Tauri (and a theorical way to support something like rspc without schema lockstep) for supporting:
 - [ ] Support for numbers larger than the `number` JS type can handle.
 - [ ] `NaN` and `Infinity`/`-Infinity` - Based on [this Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=1bac88f53162156c389c973c5f0b4ecf) and the [u32](https://doc.rust-lang.org/std/primitive.f32.html) documentation.

**Comment by oscartbeaumont** [comment]
Maybe something like:
```rust
impl DataType {
    /// TODO
    pub fn is<T: Type>(&self) -> bool {
        todo!();
    }
}
```

However, idk how that works for something like `chrono`/`uuid` which uses `PrimitiveDataType` instead of their own types. Maybe we should post-process them instead of pre-processing stuff.

**Comment by oscartbeaumont** [comment]
We also want to see if we can better express intent. Eg. date string vs string.

Usingf `DataType::is` or something similar is cool but it would require a framework like Tauri Specta to depend on *every* date library which seems counterproductive.

**Comment by oscartbeaumont** [comment]
Prototyping in #419 and #458

Related to #46, #440

**Comment by oscartbeaumont** [comment]
For future application or framework authors below I have outlined each class of these types and how to handle it:

If you want to support any of these you must start with implementing [`specta_typescript::semantic`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/index.html) into your framework.

Semantic ties together the runtime and types into a set of rules which apply to both. For example say we have `getResponse()`, this feature is able to emit something like:
```ts
getResponse().then((v) => ({  ...v, my_struct: { ...v.my_struct, date_field: new Date(v.my_struct.date_field) } }))
```
this means we can have the serialized type transform back into the proper JS runtime type while maintaining JSON and the regular serialization in the middle. This has to apply to both arguments and responses from JS as we need to apply the mapping in both directions.

Semantic is build in rules, we have built in rules for common crates but you can also define your own. Refer to the [module documentation](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/index.html) for more.

## BigInt and Floats

If you want to support `BigInt` you need both runtime and type changes.

Some facts which are worth knowing before we go ahead:
 - `NaN`, `Infinity`, `-Infinity` are serialized to `null` in Javascript and Serde JSON.
 - `BigInt` cannot hold a float or `NaN`, `Infinity`, `-Infinity`. Luckily JS can express all floating point numbers from Rust except for `f128` which is just flat out unrepresentable in JS so we don't need to worry about it.

Runtime behaviour required:
 - `NaN`/`Infinity`/`-Infinity` become `null` in [`serde_json`](https://docs.rs/serde_json) and [`JSON.*`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON) in JS, however we need to preserve these. For this to work we have no choice but to wrap the `serde_json::ser::Serializer` as early as possible. Once the JSON is transformed to `serde_json::Value` or a JSON `String` the information is lost.
 - For numbers outside the `Number.MAX_SAFE_INTEGER` and `Number.MIN_SAFE_INTEGER` range we can rely on the [`reviver`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#reviver) parameter to capture them as `BigInt` before `JSON.parse` truncates them to `number`.
 	 - Tauri doesn't always use `JSON.parse` as sometimes it's more efficient for them to just dump the value into the webview through `eval`. This means with Tauri in practice you will always want to stringify the numbers outside the safe range.

Then runtime side can be solved by using [jsone](https://docs.rs/jsone) a small crate I made for this exact purpose. You can wrap your value with `jsone::Jsone(SomeStruct { .. })` and it will use a special string encoding for bigint or floating point types when being serialized where required to ensure they are lossless. This the string value can be decoded with some helpers on the other end. This can be paired with [`specta_typescript::semantic::Configuration::enable_lossless_bigints`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/struct.Configuration.html#method.enable_lossless_bigints) and [`specta_typescript::semantic::Configuration::enable_lossless_floats`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/struct.Configuration.html#method.enable_lossless_floats) to opt-into the required type transformation to make this work properly.

With Tauri Specta/TauRPC this could be manually applied to each command, event and channel but that is annoying and easy to miss. I tried to work on an official [Tauri fork](https://github.com/oscartbeaumont/tauri/tree/jsone) which integrates `jsone` but you run across the following issues so I am not sure in practice this is solvable without massive work from Tauri:
 - Tauri allows JSON to come into the system from outside in a lot of places.
    - For example [`Emitter::emit_str`](https://docs.rs/tauri/latest/tauri/trait.Emitter.html#method.emit_str), [`Emitter::emit_str_to`](https://docs.rs/tauri/latest/tauri/trait.Emitter.html#method.emit_str_to), [`InvokeResponseBody::Json`](https://docs.rs/tauri/latest/tauri/ipc/enum.InvokeResponseBody.html#variant.Json), [`Event::payload`](https://docs.rs/tauri/latest/tauri/struct.Event.html#method.payload), [`InvokeMessage::payload`](https://docs.rs/tauri/latest/tauri/ipc/struct.InvokeMessage.html#method.payload) and probally more I didn't run across.
    - Like for example current Tauri Specta versions use `serde_json::from_string` for the raw event payload, which would bypass any work done in Tauri. It's hard for all cases to be covered in Tauri when the raw JSON APIs exist.
  - Upstreaming any implementation is basically impossible because:
      - it would have to be enabled for all IPC messages (not opt-in). As for commands we would need to put `jsone` within the [`impl<T: Serialize> IpcResponse for T {`](https://github.com/tauri-apps/tauri/blob/5e3126ff7045aec54811b227cb4d33d78b3957b5/crates/tauri/src/ipc/mod.rs#L181) implementation which has zero context to the request (Eg. headers to determine if the bigint code should be turned). This would potentially add some useless overhead to all IPC messages, even for users who don't want it. It would also change the types which would be breaking, or require undoing the operation in JS unless the user asks for it which is again overhead for anyone not using the feature.

In practice for frameworks:
 - rspc
     - would be perfectly fine supporting this as it has pluggable serializers and completly bypasses all of Tauri's stuff.
 - TauRPC
    - as it reimplement `#[tauri::command]` it would technically be possible to inject `jsone` for commands and events but as it uses `tauri::Channel` it could not wrap channels.
    - it would require cloning a lot of traits from [here](https://github.com/tauri-apps/tauri/blob/5e3126ff7045aec54811b227cb4d33d78b3957b5/crates/tauri/src/ipc/command.rs#L184) to properly wrap the users return type in `Jsone` which is a decent undertaking (but given the first issue of it being very easy to bypass, i'm not sure it's worth the maintaince burdon at the end of the day).
 - Tauri Specta
    - unable to do anything as we are by design a minimal binding to Tauri and hence leave as much to Tauri, including requiring `#[tauri::command]` to be done by the end-user so we can't control it.

All this comes together to say this is not able to be widely supported at the moment but a carefully designed RPC framework like `rspc` could make it work.

By default Specta Typescript forbids exporting bigint-style types as it shys on the side of correctness. If you want to make these export you can look into using  [`specta_util::Remapper`](https://docs.rs/specta-util/latest/specta_util/struct.Remapper.html) to remap the types similar to the following:
```rs
let mut remapper = specta_util::Remapper::new();
for primitive in [
    Primitive::usize,
    Primitive::isize,
    Primitive::u64,
    Primitive::i64,
    Primitive::u128,
    Primitive::i128,
] {
    remapper = remapper.rule(DataType::Primitive(primitive), DataType::Primitive(Primitive::u32));
}

let types = remapper.remap_types(types)

// then use `types` for exporting.
//
// Note that if your using `specta_typescript::primitives` you must also use `Remapper:remap_dt`.
```
This is provided by [`tauri_specta::Builder::dangerously_cast_bigints_to_number`](#todo) if your using that.

tldr: If you can achieve a lossless RPC layer for floats and bigints, just [`enable_lossless_bigints`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/struct.Configuration.html#method.enable_lossless_bigints) and [`enable_lossless_floats`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/struct.Configuration.html#method.enable_lossless_floats) and your off the the races!

## `Date`, `Uint8Array`, `URL`

Some facts worth knowing:
 - `JSON.stringify(new Uint8Array([1, 2]))` is `{"0":1,"1":2}` so we must do something like `[...myUint8Array]` before serialization.
 - We need to be careful with timezones when converting with the `Date` type.
 - All types need to be converted from their JSON representation (`string`) into their JS representation. Usually by something like `new Date(v)` or `new URL(v)`.

These are enabled by default when doing [`semantic::Configuration::default()`](https://docs.rs/specta-typescript/latest/specta_typescript/semantic/struct.Configuration.html#impl-Default-for-Configuration) so you don't need any special serialization framework.

If you have any questions message me or [join the Discord](https://discord.com/invite/JgqH8b4ycw) and ask! I would love to help you out!
